mlflow/store/job/abstract_store.py

```diff
@@ -0,0 +1,136 @@
+
+
+@developer_stable
+class AbstractJobStore:
+    """
+    Abstract class that defines API interfaces for storing Job metadata.
+    """
+
+    __metaclass__ = ABCMeta
+
+    def __init__(self):
+        """
+        Empty constructor. This is deliberately not marked as abstract, else every derived class
+        would be forced to create one.
+        """
+
+    @abstractmethod
+    def create_job(self, function_fullname: str, params: str, timeout: int | None = None) -> Job:
+        """
+        Create a new job with the specified function and parameters.
+
+        Args:
+            function_fullname: The full name of the function to execute
+            params: The job parameters as a string
+            timeout: The job execution timeout in seconds
+
+        Returns:
+            Job entity instance
+        """
+
+    @abstractmethod
+    def start_job(self, job_id: str) -> None:
+        """
+        Start a job by setting its status to RUNNING.
+
+        Args:
+            job_id: The ID of the job to start
+        """
+
+    @abstractmethod
+    def reset_job(self, job_id: str) -> None:
+        """
+        Reset a job by setting its status to PENDING.
+
+        Args:
+            job_id: The ID of the job to re-enqueue.
+        """
+
+    @abstractmethod
+    def finish_job(self, job_id: str, result: str) -> None:
+        """
+        Finish a job by setting its status to DONE and setting the result.
+
+        Args:
+            job_id: The ID of the job to finish
+            result: The job result as a string
+        """
+
+    @abstractmethod
+    def set_job_timeout(self, job_id: str) -> None:
+        """
+        Set a job status to Timeout.
+
+        Args:
+            job_id: The ID of the job
+        """
+
+    @abstractmethod
+    def fail_job(self, job_id: str, error: str) -> None:
+        """
+        Fail a job by setting its status to FAILED and setting the error message.
+
+        Args:
+            job_id: The ID of the job to fail
+            error: The error message as a string
+        """
+
+    @abstractmethod
+    def retry_or_fail_job(self, job_id: str, error: str) -> int | None:
+        """
+        If the job retry_count is less than maximum allowed retry count,
+        increase the retry_count and reset the job to PENDING status,
+        otherwise set the job to FAIL status and fill the job's error field.
+
+        Args:
+            job_id: The ID of the job to fail
+            error: The error message as a string
+
+        Returns:
+            If the job is allowed to retry, returns the retry count,
+            otherwise returns None.
+        """
+
+    @abstractmethod
+    def list_jobs(
+        self,
+        function_fullname: str | None = None,
+        status_list: list[JobStatus] | None = None,
+        begin_timestamp: int | None = None,
+        end_timestamp: int | None = None,
+        page_size: int = 1000,
+    ) -> Iterator[Job]:
```

Comment:
why iterator? how does this work?

Reply:
btw do we really need pagination?

Reply:
How can I fetgh the next page?

Reply:
"fetch the next page" is not supported, currently `list_jobs` is only internally used by `_enqueue_unfinished_jobs` so "fetch the next page" is not needed.  The `Iterator[job]` internally will fetch data page by page so it won't consume large memory.

Reply:
Got it. Let's remove `page_size`. It's confusing.

Reply:
But internally it still makes query from database by page. (so we need a configure to tune the page size)

Reply:
I think we can just hardcode the page size (e.g. 100).